New Weather Application - See README.md

bfwetmore opened the pull request:
# Weather Application

Uses openweathermap.org API to get a weather update via console.

bfwetmore committed:
New Weather Application - See README.md

bfwetmore committed:
Add extra line on example.env

do-work commented on README.md:
````diff
@@ -1 +1,33 @@
+
+## Install (MAC)
+
+* Install Homebrew ```/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"```
+* Install the most current version of node.js ```brew install node```
````
I like that you included this much detail.
I think its superfluous though.

If there are general things needed for the project sometimes I will include a section under install called requirements.

I like that you included a step to set up your env file.

Installing dependencies for a user of the app will hook into the `package.json` file.
As a user I should not be installing the specific dependencies. I should just be installing everything the project needs.
i.e. `npm install`

bfwetmore replied:
Done.

do-work replied:
I would include a line for `npm install` btw.

bfwetmore replied:
Done

do-work commented on README.md:
````diff
@@ -1 +1,33 @@
+
+Have your terminal opened to the weather directory.
+
+```zsh
````
Id change this to bash. Its more 'general'.

bfwetmore replied:
Done

do-work commented on README.md:
````diff
@@ -1 +1,33 @@
+Have your terminal opened to the weather directory.
+
+```zsh
+node app.js {your-city},{your-state}
````
What if someone omits a comma?
I know that is probably outside the scope of your project itself but it is something to think about especially whenever you are handling user input. In this context your user is a developer so its reasonable to expect them to follow the instructions but if the intended user was less technical you would want lots of validation on the user input.

bfwetmore replied:
Would validation be setting variables to correct the format in a variety of situations?

bfwetmore replied:
Would this be done with a conditional for each circumstance?

do-work replied:
Yea just making sure that the format is expected, adjusting it to some degree. 
Throwing an exception if the format is bad, etc, etc.

do-work commented on README.md:
````diff
@@ -1 +1,33 @@
+node app.js {zip-code}
+```
+
+## City and State Format
````
I would next this under the usage section.

bfwetmore replied:
Done.

do-work commented on app.js:
```diff
@@ -0,0 +1,38 @@
+const axios = require("axios");
+const dotenv = require('dotenv');
+dotenv.config();
```
Check JS conventions.

I thought there is supposed to be a blank line between requirements and actual code.

bfwetmore replied:
After browsing popular repos, you are right about that. Done.

do-work commented on app.js:
```diff
@@ -0,0 +1,38 @@
+dotenv.config();
+const apiURLPrefix = "http://api.openweathermap.org/";
+
+async function PrintWeather(input) {
```
Why is this function capitalized?

All the functions need a Docblock.

This method is called `printWeather` but it is also making an API request. This feels unnatural to me.
Lets split this up.
Get the weatherData in one method, add a method that builds the output, then handle the output where that is used.

Happy to talk about that if its confusing.

bfwetmore replied:
I think I got it, let me know what you think.

do-work commented on app.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+let entry = process.argv.slice(2).toString();
+userDataInput(entry);
```
Since this is the 'app' itself having the function name be `userDataInput `feels like one step of many.
I would rather see a method called `getWeather()`

bfwetmore replied:
Done.

do-work commented on app.js:
```diff
@@ -0,0 +1,38 @@
+
+async function userDataInput(entry) {
+  if (isNaN(entry)) {
+    const geoURL = `${apiURLPrefix}geo/1.0/direct?q=${entry},us&appid=${process.env.W_API}`;
+    let data = await buildResponseBody(geoURL);
+    if (data.length === 0) {
+      return console.error(`${entry} is not a valid location, please try again.`);
+    }
```
Id like to see this in its own method called `getCoordinates` or some variation of that.

bfwetmore replied:
Done.

bfwetmore committed:
Adjustments per change requests. Also makes additional adjustments to layout of the code for readability.

do-work commented on app.js:
```diff
@@ -1,38 +1,3 @@
-let entry = process.argv.slice(2).toString();
-userDataInput(entry);
+const weather = require("./weather");
+const input = process.argv.slice(2).toString();
```
Add a blank line between your require and your actual code.
I really like how clean this entry point is btw.

bfwetmore replied:
Done

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+ * @param {string|number}input
+ * @returns {Promise<void>}
+ */
+async function getWeather(input) {
```
I like the docblock to kind of match up with the function name.

So I would say something like
"Get weather data based on zipcode or address."

bfwetmore replied:
Done

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+ * @returns {Promise<void>}
+ */
+async function getWeather(input) {
+    if (isNaN(input)) {
```
Yea I think the first step here should be some validation (in another method). Throw an exception if the input is bad. Catch that exception and decide what output you want.

bfwetmore replied:
I added a function here to check for spaces in the formatting. I could go down a rabbit hole of doing additional checks of correct formatting. But I figured this would show that I can get there, and move on. Unless you think it's worth adding a bunch more, or if there is a simpler approach that I am missing?

do-work replied:
This is good.
Its just something to always think about.

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+ */
+async function getWeather(input) {
+    if (isNaN(input)) {
+        const data = await getCoordinates(input);
```
Variable names like 'data' are not good.
How about something like `coordinates` ?

bfwetmore replied:
Good catch, much better name.

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+async function getWeather(input) {
+    if (isNaN(input)) {
+        const data = await getCoordinates(input);
+        await getWeatherData(`lat=${data[0]["lat"]}&lon=${data[0]["lon"]}`);
+        return;
```
Can you do `return await ...` ?

bfwetmore replied:
Yes, Done.

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+    const geoURL = `${apiURLPrefix}geo/1.0/direct?q=${input},us&appid=${process.env.W_API}`;
+    let data = await buildResponseBody(geoURL);
+    if (data.length === 0) {
+        return console.error(`${input} is not a valid location, please try again.`);
```
Throw an exception rather than `console.error` here.
You can catch that exception where this method is used and handle it accordingly (console.error at that point).
That allows this method to be more reusable accounting for various context specific ways of handling things.

bfwetmore replied:
Done. Let me know what you think.

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+        const request = await axios.get(url);
+        responseBody = request.data;
+    } catch (error) {
+        console.error('There was an error with you entry, please see the "README" file to see correct formatting.');
```
Throw an exception.
Maybe try a custom one.

bfwetmore replied:
Done

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+ */
+async function getWeatherData(input) {
+    let url = `${apiURLPrefix}data/2.5/weather?${input}&units=imperial&appid=${process.env.W_API}`;
+    const weatherJSON = await buildResponseBody(url);
+    printWeather(weatherJSON);
```
Your getting the data AND handling the data here.
return `weatherJSON `and then you can `printWeather` where this method is called from.

bfwetmore replied:
Done.

do-work commented on weather.js:
```diff
@@ -0,0 +1,70 @@
+async function buildResponseBody(url) {
+    let responseBody = "";
+    try {
+        const request = await axios.get(url);
+        responseBody = request.data;
```
Try/catches _should_ be as small and succinct as possible.
The error you are catching is only going to happen in the `axios.get` call so pull line 55 out of the try catch and put it below.

bfwetmore replied:
Found the request variable unnecessary anyways. Tied it all back into the one responseBody variable.

bfwetmore committed:
Adjustments per change requests.

More pointed exceptions and error messages. Leaving reusable functions from being tightly coupled with specific errors.

Add simple function for formatting errors. Can be built on to add more specific format errors.

do-work commented on weather.js:
```diff
@@ -5,17 +5,31 @@ dotenv.config();
+ */
+function formatCheck(input) {
+    if (input.indexOf(',') >= 2) {
+        throw 'format error';
```
We want a custom exception - look at this - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw#throw_an_object

do-work commented on weather.js:
```diff
@@ -5,17 +5,31 @@ dotenv.config();
+ * Checks user input format
+ * @param {string|number}input
+ */
+function formatCheck(input) {
```
I like to name methods that throw an exception 'validateXYZ'.
The validation implies that an exception will be thrown. Also the doblock should contain that.

When I see a method named `formatCheck` I think that its going to DO some formatting.
Sometimes I see methods called `checkInput` in which case I think its going to return a boolean.

Its all about having a descriptive verb and then adapting based on your team's general usage.

do-work commented on weather.js:
```diff
@@ -5,17 +5,31 @@ dotenv.config();
-        return;
+    try {
+        formatCheck(input);
+        if (isNaN(input)) {
+            const data = await getCoordinates(input);
+            return printWeather(await getWeatherData(`lat=${data[0]["lat"]}&lon=${data[0]["lon"]}`));
+        }
+        printWeather(await getWeatherData(`zip=${input}`));
```
I like all this.
Looks very good to me.
Its succinct without in-lining too much and its clear what each step does.

do-work commented on weather.js:
```diff
@@ -25,11 +39,11 @@ async function getWeather(input) {
-    if (data.length === 0) {
-        return console.error(`${input} is not a valid location, please try again.`);
+    try {
+        return await buildResponseBody(geoURL);
```
This feels weird.

When I see 'buildResponseBody' I think that the request was already made.
Having that method DO the request feels wrong.
I would do the request in this method and then hand off the response object to `buildResponseBody `and rename it something like `parseResponse` or something.

do-work reviewed (changes requested):
We should pair on a couple details but its almost done. 
Good job pushing through it.
